Sources/OpenAPIRuntime/EventStreams/ServerSentEventsDecoding.swift

```diff
@@ -0,0 +1,451 @@
+        enum NextAction {
+
+            /// Return nil to the caller, no more bytes.
+            case returnNil
+
+            /// Emit a completed event.
+            case emitEvent(ServerSentEvent)
+
+            /// The line is not complete yet, needs more bytes.
+            case needsMore
+
+            /// Rerun the parsing loop.
+            case noop
+        }
+
+        /// Read the next line parsed from upstream bytes.
+        /// - Returns: An action to perform.
+        mutating func next() -> NextAction {
+            switch state {
+            case .accumulatingEvent(var event, var buffer):
+                guard let line = buffer.first else { return .needsMore }
+                state = .mutating
+                buffer.removeFirst()
+                if line.isEmpty {
+                    // Dispatch the accumulated event.
+                    state = .accumulatingEvent(.init(), buffer: buffer)
+                    // If the last character of data is a newline, strip it.
+                    if event.data?.hasSuffix("\n") ?? false { event.data?.removeLast() }
+                    return .emitEvent(event)
+                }
+                if line.first! == ASCII.colon {
+                    // A comment, skip this line.
+                    state = .accumulatingEvent(event, buffer: buffer)
+                    return .noop
+                }
+                // Parse the field name and value.
+                let field: String
+                let value: String?
+                if let indexOfFirstColon = line.firstIndex(of: ASCII.colon) {
+                    field = String(decoding: line[..<indexOfFirstColon], as: UTF8.self)
+                    let valueBytes = line[line.index(after: indexOfFirstColon)...]
+                    let resolvedValueBytes: ArraySlice<UInt8>
+                    if valueBytes.isEmpty {
+                        resolvedValueBytes = []
+                    } else if valueBytes.first! == ASCII.space {
+                        resolvedValueBytes = valueBytes.dropFirst()
+                    } else {
+                        resolvedValueBytes = valueBytes
+                    }
+                    value = String(decoding: resolvedValueBytes, as: UTF8.self)
+                } else {
+                    field = String(decoding: line, as: UTF8.self)
+                    value = nil
+                }
+                guard let value else {
+                    // An unknown type of event, skip.
+                    state = .accumulatingEvent(event, buffer: buffer)
+                    return .noop
+                }
+                // Process the field.
+                switch field {
+                case "event": event.event = value
+                case "data":
+                    var data = event.data ?? ""
+                    data.append(value)
+                    data.append("\n")
+                    event.data = data
+                case "id": event.id = value
```

Comment:
Spec says:
> If the field value does not contain U+0000 NULL, then set the last event ID buffer to the field value. Otherwise, ignore the field.

Are we relying on earlier logic / string parsing for this?

Reply:
The specification describes the EventSource abstraction, a stateful representation of a stream, which helps you work with Server-sent Events.

My implementation only provides the Server-sent Events portion, but not the EventSource part. 

EventSource is currently up to adopters to implement to their liking, because it also contains actually making requests that establish the stream, which is out of scope here. In the future we could also implement EventSource it in the runtime library, but it wasn't a blocker here, which is why it was kept out.

The EventSource value has a "last event ID" field, so it's not relevant to the underlying SSE encoder/decoder.

Reply:
OK, thanks for confirming!